### PR TITLE
Network performance

### PR DESCRIPTION
**Pull request by marseel**
Introduce network performance testing in CI.

Commit messages copied for convenience:

```
    Add network performance CI test.
    
    For now, we cover following matrix of features:
    - tunneling/direct-routing
    - no encryption/ipsec
    - hubble enabled/disabled
    
    All results are exported in a format compatible with Perfdash, where we
    can visualize results and see regressions/improvements for specific
    configurations.
```

Example run: https://github.com/cilium/cilium/actions/runs/7612606888

### CODE REVIEW

File: .github/workflows/net-perf-gke.yaml [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+          echo sha=${SHA} >> $GITHUB_OUTPUT
+          echo owner=${OWNER} >> $GITHUB_OUTPUT
+
+      # Warning: since this is a privileged workflow, subsequent workflow job
+      # steps must take care not to execute untrusted code.
+      - name: Checkout pull request branch (NOT TRUSTED)
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          ref: ${{ steps.vars.outputs.sha }}
+          persist-credentials: false
```
Comment (giorio94) [inline .github/workflows/net-perf-gke.yaml]: I'd suggest dropping this step given that this workflow is executed on schedule only.

**Review by giorio94 — COMMENTED** [review]
Disclaimer: I haven't reviewed the full PR.

Just one inline comment about checking out the untrusted context, which is not necessary in this context (I was actually to suggest making it consistent with the changes in https://github.com/cilium/cilium/pull/30207, but it turns out it is not even necessary).

File: .github/actions/set-env-variables/action.yml [hunk 2]
```diff
@@ -15,3 +15,4 @@ runs:
         CILIUM_CLI_VERSION="v0.15.20"
         echo "CILIUM_CLI_VERSION=$CILIUM_CLI_VERSION" >> $GITHUB_ENV
         echo "PUSH_TO_DOCKER_HUB=true" >> $GITHUB_ENV
+        echo "GCP_PERF_RESULTS_BUCKET=gs://cilium-perf-results" >> $GITHUB_ENV
```
Comment (learnitall) [inline .github/actions/set-env-variables/action.yml]: Can we remove this variable from the other scale workflows? They still set it manually: https://github.com/cilium/cilium/blob/7fcdb47469dc4c7f4326284404913efecad6b065/.github/workflows/scale-test-100-gce.yaml#L39

File: .github/workflows/net-perf-gke.yaml [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    ${{ github.event_name }}
+    ${{
+      (github.event_name == 'schedule' && github.sha) ||
+      (github.event_name == 'workflow_dispatch' && github.event.inputs.PR-number)
+    }}
+  cancel-in-progress: true
+
+env:
+  clusterName: ${{ github.repository_owner }}-${{ github.event.repository.name }}-${{ github.run_id }}-${{ github.run_attempt }}
+  cilium_cli_ci_version:
```
Comment (learnitall) [inline .github/workflows/net-perf-gke.yaml:35]: This variable is blank, is that intentional?

File: .github/workflows/net-perf-gke.yaml [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+      - name: Install gke-gcloud-auth-plugin
+        run: |
+          gcloud components install gke-gcloud-auth-plugin
+
+      - name: Display gcloud CLI info
+        run: |
+          gcloud info
+
+      - name: Create GKE cluster
```
Comment (learnitall) [inline .github/workflows/net-perf-gke.yaml:186]: Is there a reason for using the gcloud cli, rather than the [scale test action](https://github.com/cilium/scale-tests-action/blob/main/create-cluster/action.yaml) that we have?

File: .github/workflows/net-perf-gke.yaml [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            --machine-type n2-standard-2 \
+            --disk-type pd-standard \
+            --disk-size 20GB \
+            --node-taints ignore-taint.cluster-autoscaler.kubernetes.io/cilium-agent-not-ready=true:NoExecute
+
+      - name: Get cluster credentials
+        run: |
+          gcloud container clusters get-credentials ${{ env.clusterName }}-${{ matrix.index }} --zone ${{ env.gcp_zone }}
+
+      - name: Wait for images to be available
```
Comment (learnitall) [inline .github/workflows/net-perf-gke.yaml:207]: I'd recommend moving this step before cluster creation. IMO it makes sense to confirm the images are available before going through the trouble of creating a cluster. If the images aren't available for whatever reason, say quay is down, then we don't have to create the cluster unnecessarily.

File: .github/workflows/net-perf-gke.yaml [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 40
+    env:
+      job_name: "Installation and Perf Test"
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - index: 1
+            name: "native"
```
Comment (learnitall) [inline .github/workflows/net-perf-gke.yaml:55]: Optional nit: for readability, it would be nice if all variables are included, even if they aren't used. IMO it would be easier to see it all explicitly spelled-out, rather than having to keep it in my head.

```suggestion
          - index: 1
            name: "native"
            mode: "native"
            encryption: "none"
            hubble: "false"
```

File: .github/workflows/net-perf-gke.yaml [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+  check_url: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
+  USE_GKE_GCLOUD_AUTH_PLUGIN: True
+  gcp_zone: us-east5-a
+  k8s_version: 1.28
+
+jobs:
+  installation-and-perf:
+    name: Installation and Perf Test
+    runs-on: ubuntu-latest
+    timeout-minutes: 40
```
Comment (learnitall) [inline .github/workflows/net-perf-gke.yaml]: Maybe we should bump this to 60? The wait for images step has a timeout of 30 minutes, meaning this line implies that all the other steps should take no more than 10 minutes. In a worst-case scenario where waiting for images takes 20 minutes, having some extra buffer room would be nice.

File: .github/workflows/net-perf-gke.yaml [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        run: |
+          cilium status --wait --wait-duration=10m
+          kubectl get pods -n kube-system
+          kubectl -n kube-system exec daemonset/cilium -- cilium-dbg status
+
+      - name: Run perf test (${{ matrix.name }})
+        id: run-perf
+        run: |
+          mkdir output
+          cilium connectivity perf --duration=30s --host-net=true --pod-net=true --report-dir=./output
```
Comment (learnitall) [inline .github/workflows/net-perf-gke.yaml:236]: I'd recommend bumping this duration up a bit, which will help tone down the variance. Three minutes would be ideal, but we should at least do one minute:

```suggestion
          cilium connectivity perf --duration=60s --host-net=true --pod-net=true --report-dir=./output
```

**Review by learnitall — CHANGES_REQUESTED** [review]
Exciting stuff! I have a couple of questions, but otherwise LGTM.

File: .github/workflows/net-perf-gke.yaml (hunk 3 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:35]: Yes, that's what all other workflows do too. Use-case is that you can specify SHA of cilium-cli built before it's released for testing for example.

File: .github/workflows/net-perf-gke.yaml (hunk 4 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:186]: For now, I don't see a use-case for it. I chose GKE cluster so we don't need to manage control plane as k8s control plane shouldn't really matter for network performance.

File: .github/workflows/net-perf-gke.yaml (hunk 1 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml]: Done.

File: .github/workflows/net-perf-gke.yaml (hunk 5 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:207]: On the other hand, I think it's safe to assume that quay works most of the time and we can built images on main branch - in that case having order of creating cluster -> getting images lowers total duration of workflow.

File: .github/workflows/net-perf-gke.yaml (hunk 6 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:55]: Makes sense, done!

File: .github/workflows/net-perf-gke.yaml (hunk 7 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml]: Sounds good. Bumped to 60.
FYI currently workflow takes ~30 minutes in total of which performing the test is 20 minutes.

File: .github/actions/set-env-variables/action.yml (hunk 2 above)
Comment (marseel) [reply .github/actions/set-env-variables/action.yml]: Good idea, done.

File: .github/workflows/net-perf-gke.yaml [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        shell: bash {0} # Disable default fail-fast behavior so that all commands run independently
+
+      - name: Export results and sysdump to GS bucket
+        if: ${{ always() && steps.run-perf.outcome != 'skipped' && steps.run-perf.outcome != 'cancelled' }}
+        uses: cilium/scale-tests-action/export-results@8a522b9d71254b6f6615c296e41a70610c9615ea # main
+        with:
+          test_name: ${{ env.test_name }}-${{ matrix.name }}
+          results_bucket: ${{ env.GCP_PERF_RESULTS_BUCKET }}
+          artifacts: ./output/*
+          other_files: cilium-sysdump-final.zip
```
Comment (bimmlerd) [inline .github/workflows/net-perf-gke.yaml]: nit: no newline at EOF

**Review by bimmlerd — APPROVED** [review]
LGTM

**Comment by marseel** [comment]
I've added backports to 1.15-1.13

File: .github/workflows/net-perf-gke.yaml (hunk 8 above)
Comment (learnitall) [reply .github/workflows/net-perf-gke.yaml:236]: Talking offline, we'll keep it at 30s and increase as needed. The goal is to try and balance this duration value and the duration of the workflow overall.

**Review by learnitall — APPROVED** [review]

File: .github/workflows/net-perf-gke.yaml [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+          if [ "${{ matrix.mode }}" = "tunnel" ] ; then
+            CILIUM_INSTALL_DEFAULTS+=" --datapath-mode=tunnel"
+          fi
+
+          if [ "${{ matrix.encryption }}" = "ipsec" ] ; then
+            CILIUM_INSTALL_DEFAULTS+=" --helm-set=encryption.enabled=true --helm-set=encryption.type=ipsec"
+          fi
+
+          echo cilium_install_defaults=${CILIUM_INSTALL_DEFAULTS} >> $GITHUB_OUTPUT
+          echo sha=${SHA} >> $GITHUB_OUTPUT
```
Comment (thorn3r) [inline .github/workflows/net-perf-gke.yaml]: I think this is a dupe of line 135

**Review by thorn3r — APPROVED** [review]
Looks great!

**Review by viktor-kurchenko — APPROVED** [review]

File: .github/workflows/net-perf-gke.yaml [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+      (github.event_name == 'workflow_dispatch' && github.event.inputs.PR-number)
+    }}
+  cancel-in-progress: true
+
+env:
+  clusterName: ${{ github.repository_owner }}-${{ github.event.repository.name }}-${{ github.run_id }}-${{ github.run_attempt }}
+  cilium_cli_ci_version:
+  CILIUM_CLI_MODE: helm
+  test_name: gke-perf
+  check_url: https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}
```
Comment (giorio94) [inline .github/workflows/net-perf-gke.yaml:38]: Nit. `check_url` seems unused and could be dropped.

File: .github/workflows/net-perf-gke.yaml [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+  USE_GKE_GCLOUD_AUTH_PLUGIN: True
+  gcp_zone: us-east5-a
+  k8s_version: 1.28
```
Comment (giorio94) [inline .github/workflows/net-perf-gke.yaml:41]: Nit. Should `k8s_version` be managed by renovate for automatic updates?

File: .github/workflows/net-perf-gke.yaml [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+            --helm-set=operator.image.suffix=-ci \
+            --helm-set=operator.image.tag=${SHA} \
+            --helm-set=operator.image.useDigest=false \
+            --helm-set=clustermesh.apiserver.image.repository=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/clustermesh-apiserver-ci \
+            --helm-set=clustermesh.apiserver.image.tag=${SHA} \
+            --helm-set=clustermesh.apiserver.image.useDigest=false \
+            --helm-set=hubble.relay.image.repository=quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/hubble-relay-ci \
+            --helm-set=hubble.relay.image.tag=${SHA} \
+            --helm-set=hubble.relay.image.useDigest=false"
+          fi
```
Comment (giorio94) [inline .github/workflows/net-perf-gke.yaml:148]: You may want to reuse the [`helm-default` action](https://github.com/cilium/cilium/blob/4fe2fd31b1230a7b00f929661f31caff23ee98a7/.github/workflows/conformance-clustermesh.yaml#L203-L208) to configure these settings. In that case, you'd need to override the monitor aggregation setting though.

File: .github/workflows/net-perf-gke.yaml [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+
+          if [ "${{ matrix.mode }}" = "tunnel" ] ; then
+            CILIUM_INSTALL_DEFAULTS+=" --datapath-mode=tunnel"
+          fi
```
Comment (giorio94) [inline .github/workflows/net-perf-gke.yaml]: Nit. I'd explicitly configure the mode in both cases, instead of relying on the CLI default, as it could lead to surprises if ever using a different cloud provider.

**Review by giorio94 — APPROVED** [review]
/lgtm

Just a couple of nits inline.

File: .github/workflows/net-perf-gke.yaml (hunk 11 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:38]: I will leave it for now - it will be useful once we can trigger this manually.

File: .github/workflows/net-perf-gke.yaml (hunk 12 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:41]: I don't think we can do that as we use GKE clusters, for conformance GKE tests we update that manually in https://github.com/cilium/cilium/blob/c8ed4875cdb59b41bea22b7fdf5eb6450a62d052/.github/actions/gke/k8s-versions.yaml#L14

File: .github/workflows/net-perf-gke.yaml (hunk 13 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml:148]: I was thinking about doing that, but then decided not to reuse common actions as it might impact results in case we change them.

File: .github/workflows/net-perf-gke.yaml (hunk 14 above)
Comment (marseel) [reply .github/workflows/net-perf-gke.yaml]: Makes sense, fixed. 

I'm not resolving this comment before rerunning tests to double check everything works.

**Comment by marseel** [comment]
Passing run with all changes: https://github.com/cilium/cilium/actions/runs/7709171650/job/21013110382

File: .github/workflows/net-perf-gke.yaml [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+        id: install-cilium
+        run: |
+          cilium install --dry-run-helm-values ${{ steps.vars.outputs.cilium_install_defaults }}
+          cilium install ${{ steps.vars.outputs.cilium_install_defaults }}
+
+      - name: Wait for Cilium to be ready
+        run: |
+          cilium status --wait --wait-duration=10m
+          kubectl get pods -n kube-system
+          kubectl -n kube-system exec daemonset/cilium -- cilium-dbg status
```
Comment (julianwiedmann) [inline .github/workflows/net-perf-gke.yaml:230]: @marseel Drive-by - is this workflow actually working on `v1.13` and `v1.14` ? We backported the `cilium-dbg` part as-is, but back then the name was different.

Comment (learnitall) [reply .github/workflows/net-perf-gke.yaml:230]: Looks like it has only been running on main: https://github.com/cilium/cilium/actions/workflows/net-perf-gke.yaml?query=

Comment (giorio94) [reply .github/workflows/net-perf-gke.yaml:230]: Scheduled workflows are run only from the default branch. 

The scheduled runs for the other Ariane-triggered workflows are run through https://github.com/cilium/cilium/blob/main/.github/workflows/ariane-scheduled.yaml, but we would need to extend that to also cover this workflow (if that's the goal, which I guess it is).
